File: trunk/dotnet/picocontainer/src/Defaults/DecoratingComponentAdapter.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using PicoContainer;
using PicoContainer.Defaults;


namespace PicoContainer.Defaults{
  public class DecoratingComponentAdapter : IComponentAdapter {
    private IComponentAdapter theDelegate;

    public DecoratingComponentAdapter(IComponentAdapter theDelegate) {
      this.theDelegate = theDelegate;
    }

    virtual public object ComponentKey {
      get {
        return theDelegate.ComponentKey;
      }
    }

    virtual public Type ComponentImplementation {
      get {
        return theDelegate.ComponentImplementation;
      }
    }

    public virtual object ComponentInstance{
      get {
        return theDelegate.ComponentInstance;
      }
    }

    public virtual void Verify() {
      theDelegate.Verify();
    }

    public virtual IComponentAdapter Delegate {
      get {
        return theDelegate;
      }			
    }

    public virtual IPicoContainer Container {
      get {
        return theDelegate.Container;
      }
      set {
        theDelegate.Container = value;
      }
    }
  }
}

File: java/picocontainer/src/java/org/picocontainer/defaults/DecoratingComponentAdapter.java
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Original code by Jon Tirsen                                               *
 *****************************************************************************/

package org.picocontainer.defaults;

import org.picocontainer.ComponentAdapter;
import org.picocontainer.PicoContainer;
import org.picocontainer.PicoInitializationException;
import org.picocontainer.PicoIntrospectionException;
import org.picocontainer.PicoVisitor;

import java.io.Serializable;

/**
 * @author Jon Tirsen (tirsen@codehaus.org)
 * @author Aslak Hellesoy
 * @version $Revision$
 */
public class DecoratingComponentAdapter implements ComponentAdapter, Serializable {

    private ComponentAdapter delegate;

    public DecoratingComponentAdapter(ComponentAdapter delegate) {
        this.delegate = delegate;
    }

    public Object getComponentKey() {
        return delegate.getComponentKey();
    }

    public Class getComponentImplementation() {
        return delegate.getComponentImplementation();
    }

    public Object getComponentInstance() throws PicoInitializationException, PicoIntrospectionException, AssignabilityRegistrationException, NotConcreteRegistrationException {
        return delegate.getComponentInstance();
    }

    public void verify() {
        delegate.verify();
    }

    public PicoContainer getContainer() {
        return delegate.getContainer();
    }

    public ComponentAdapter getDelegate() {
        return delegate;
    }

    public void setContainer(PicoContainer picoContainer) {
        delegate.setContainer(picoContainer);
    }

    public void accept(PicoVisitor visitor) {
        visitor.visitComponentAdapter(this);
        delegate.accept(visitor);
    }
}
